ark/analysis/spatial_analysis.py

```diff
@@ -297,6 +297,95 @@ def create_neighborhood_matrix(all_data, dist_matrices_dict, included_fovs=None,
+                                    fov_col='SampleID', label_col='cellLabelInImage',
+                                    cell_type_col='cell_type'):
+    """Generate the cluster info on all_data based on k=cluster_num k-means clustering
+    on neighbor_mat. cluster_num has to be picked based on the visualization results generated
+    from compute_cluster_metrics.
+
+    Args:
+        all_data (pandas.DataFrame):
+            data including fovs, cell labels, and cell expression matrix for all markers
+        neighbor_mat (pandas.DataFrame):
+            a neighborhood matrix, created from create_neighborhood_matrix
+        cluster_num (int):
+            the optimal k to pass into k-means clustering to generate the final clusters
+            and corresponding results
+        excluded_colnames (list):
+            all column names that are not markers. If argument is none, default is
+            ["cell_size", "Background", "HH3",
+            "summed_channel", "label", "area",
+            "eccentricity", "major_axis_length",
+            "minor_axis_length", "perimeter", "fov"]
+        included_fovs (list):
+            patient labels to include in analysis. If argument is None, default is all labels used.
+        cluster_label_col (str):
+            the name of the cluster label col we will create
+        fov_col (str):
+            the name of the column in all_data and neighbor_mat determining the fov
+        label_col (str):
+            the name of the column in all_data and neighbor_mat determining the label
+        cell_type_col (str):
+            the name of the column in all_data determining the cell type
+
+    Returns:
+        tuple (pandas.DataFrame, pandas.DataFrame):
+
+        - a matrix listing the number of cell types per cluster
+        - a matrix listing the mean expression of each marker per cluster
+    """
+
+    # error checking
+    if included_fovs is None:
+        included_fovs = neighbor_mat[fov_col].unique()
+
+    if excluded_colnames is None:
+        excluded_colnames = ["cell_size", "Background", "HH3",
+                             "summed_channel", "label", "area",
+                             "eccentricity", "major_axis_length", "minor_axis_length",
+                             "perimeter", "fov"]
+
+    # make sure the specified
+    if not np.isin(excluded_colnames, all_data.columns).all():
+        raise ValueError("Column names were not found in Expression Matrix")
+
+    if not np.isin(included_fovs, neighbor_mat[fov_col]).all():
+        raise ValueError("Not all specified fovs exist in the provided neighborhood matrix")
+
+    if cluster_num < 2:
+        raise ValueError("Invalid k provided for clustering")
+
+    # subset neighbor mat
+    neighbor_mat_data = neighbor_mat[neighbor_mat[fov_col].isin(included_fovs)]
+
+    # drop non-channel columns
+    col_to_drop = [fov_col, label_col]
+    neighbor_mat_data = neighbor_mat_data.drop(col_to_drop, axis=1)
```

Comment:
We'll need to do some error checking to make sure remaining columns are only numeric

Reply:
I'm not sure if this will work because `label_col` will also be a numeric type, but we don't want that in the clustering process. I think we should simply include a cell in the `Deepcell_Postprocessing` notebook that handles dropping these columns prior to passing them into `compute_cluster_metrics`. I will adjust the tests accordingly.